#############################################################################
#          Rules to generate coverage plots.                                #
#                                                                           #
# Date: 02.01.2015                                                          #
#                                                                           #
#############################################################################

from snakemake.utils import R

ruleorder: plot_hist > plot_regions

rule plot_regions:
    """Produces coverage plots for each gene in the manifest file."""
    input: expand("results/coverage/{{mapper}}/{{ref}}/{{kind}}/{sample}_" + config["platform"] + "_" + config["manifest"] + "_20_20.cov", sample = config["samples"].keys() )
    output: "plots/coverage/{mapper}/{ref}/{kind}/{anyregion}.pdf"
    run: 
      files = "c({})".format(",".join(map( lambda str: '"' + str + '"', input))),
      mapper = wildcards.mapper,
      ref = wildcards.ref,
      kind = wildcards.kind,
      platform = config["platform"],
      manifest = config["manifest"],
      left = "20",
      right = "20",
      run = config["run"],
      R("""
      library(ggplot2)
      library(data.table)
      library(gridExtra)

      files <- {files}
      labels <- sub( '_.*', "", basename(files)) # extract sample names
      coverages<-list()
      for (i in seq_along(files)) {{
        coverages[[i]] <- fread(files[i], sep="\t")
        coverages[[i]] <- cbind( labels[i], coverages[[i]])
        setnames(coverages[[i]], c("Sample", "Chromosome", "Start", "End", "FeatureRegionName", "Position", "Coverage"))
      }}

      dt <- do.call('rbind', coverages)
      dt[,RegionName := sub(".*?-", "", dt[,FeatureRegionName])]
      dt[,Region := as.numeric(sub("-.*", "", dt[,RegionName]))]
      dt[,Feature := sub("-.*", "", dt[,FeatureRegionName])]
      dt[,Name := sub(".*-", "", dt[,FeatureRegionName])]
      dt[,Position := dt[,Position - {left}]]
      setkey(dt, Feature)

      setwd('plots/coverage/{mapper}/{ref}/{kind}/')
      
      for (f in unique( dt[,Feature] ) ) {{
      
        cat(paste("Plotting", dt[f,Name][1], f))
        features <- split(dt[f,], f=dt[f,Region])

        nd<-function(x, top=10){{return(ceiling(x/ceiling(x/top)))}}

        features2 <- list()

        plots_per_page <- nd(length(features))

        for (i in seq(1,length(features), plots_per_page)) {{
          features2 <- c(features2, list(do.call(rbind, features[i:min(i+plots_per_page-1, length(features))])))
        }}

        plots<-lapply(features2, function(features){{
              p <- ggplot( data = features, aes( x = Position, y = Coverage, colour = Sample ))
              p <- p + geom_line() + geom_hline( aes(yintercept = 100)) 

              p <- p + geom_vline( xintercept = 0.5, colour="grey", linetype = "longdash" )
              p <- p + geom_vline( aes( xintercept = End - Start - {left} - {right} + 0.5 ), colour="grey", linetype = "longdash" )

              p <- p + facet_wrap( ~ Region, scales = 'free_x', ncol=1 )
              
              p <- p + labs(title = paste0('Coverage of ', f, '(', features[1,Name], ') on ', features[1,Chromosome], ' in {ref}, run: {run}'), x='Position in Exon' )
              p <- p + theme(plot.title = element_text(size = rel(0.8)))
              return(p)
        }})

        pdf(paste0(dt[f,Name][1], '-', f, ".pdf"), paper="a4", height=11.692, width=8.267)
          lapply(plots, print)
          lapply(plots, function(p){{
            print(p + scale_y_log10(limits = c(1, 20000), breaks = 10^(1:4))
          )}}) 
        dev.off()
      }}
      """)


rule plot_hist:
    """Produces a summary plot of target coverage."""
    input: expand("results/coverage/{{mapper}}/{{ref}}/{{kind}}/{sample}_" + config["platform"] + "_" + config["manifest"] + "_20_20.hist", sample = config["samples"].keys() )
    output: "plots/coverage/{mapper}/{ref}/{kind}/Total.pdf"
    priority: 20
    run:
      files = "c({})".format(",".join(map( lambda str: '"' + str + '"', input))),
      mapper = wildcards.mapper,
      ref = wildcards.ref,
      kind = wildcards.kind,
      platform = config["platform"],
      manifest = config["manifest"],
      left = "20",
      right = "20",
      run = config["run"],
      R("""
      library(ggplot2)

      files <- {files}
      labels <- sub( '_.*', "", files) # extract sample names
      cov <- list()

      for (i in seq_along(files)) {{
      cov[[i]] <- read.table(files[i], sep="\t")
      cov[[i]] <- cov[[i]][,c(-1, -4)] # drop first column
      cov[[i]] <- cbind( cov[[i]], 1 - cumsum( cov[[i]][,3] ) ) # add cummulative frequency of depth
      cov[[i]] <- cbind( labels[i], cov[[i]]) # add label column
      }}
      
      df <- do.call('rbind', cov)
      colnames(df) <- c('Sample', 'Depth', 'Count', 'Frequency', 'CumFrequency')

      setwd('plots/coverage/{mapper}/{ref}/{kind}/')

      p <- ggplot( data = df, aes( x = Depth, y = CumFrequency, colour = Sample ) ) 
      p <- p + geom_line() + xlim(0, 500) + ylim(0.7, 1.0)
      p <- p + geom_vline( xintercept = c(20,50,100), linetype = "longdash" )
      p <- p + ggtitle("Coverage distribution in run {run} using {ref}") + xlab("Depth of coverage") 
      p <- p + ylab( sprintf("Fraction of target bases with a coverage >= depth" ) )
      ggsave( file = "Total.pdf", plot = p, width = 10, height = 10, title = "Coverage distribution of target regions" )
      """)
