#############################################################################
#          Script to generate human reference data.                         #
#                                                                           #
# Date: 23.02.2015                                                          #
#                                                                           #
# Example usage: snakemake --snakefile human.rules                          #
#############################################################################

######### GENOMES

## genome hg38

rule genome_hg38_download:
    output: config["ref_dir"] + "/human/hg38/genome/hg38.fa"
    shell: "wget -q -O {output} ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/GRCh38_reference_genome/hs38DH.fa"

## genome hg19

rule genome_hg19_download:
    output: config["ref_dir"] + "/human/hg19/genome/hg19.fa"
    shell: "wget -q -O - ftp://ftp-trace.ncbi.nih.gov//1000genomes/ftp/technical/reference/human_g1k_v37.fasta.gz | gunzip -cd > {output}"

rule hg19_ucsc_rm_download:
    output: config["ref_dir"] + "/human/hg19/genome/hg19_ucsc_rm.fa"
    shell: "wget -q -O - --user=gsapubftp-anonymous ftp://ftp.broadinstitute.org//bundle/2.8/hg19/ucsc.hg19.fasta.gz | gunzip -cd > {output}"

######### GENE ANNOTATION

rule gene_hg38_download:
    output: config["ref_dir"] + "/human/hg38/genes/all.gtf"
    shell: "wget -q -O - ftp://ftp.ensembl.org/pub/release-80/gtf/homo_sapiens/Homo_sapiens.GRCh38.80.gtf.gz | gunzip -cd > {output}"

rule gene_hg19_download:
    output: config["ref_dir"] + "/human/hg19/genes/all.gtf"
    shell: "wget -q -O - ftp://ftp.ensembl.org/pub/release-75/gtf/homo_sapiens/Homo_sapiens.GRCh37.75.gtf.gz | gunzip -cd > {output}"

######### VARIANT DATABASES

## dbSNP hg38
rule dbsnp_hg38_download:
    output: config["ref_dir"] + "/human/hg38/snp/dbsnp142.vcf"
    shell: "wget -q -O - ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b142_GRCh38/VCF/All_20150218.vcf.gz | gunzip -cd | sed 's/^\([^#]\)/chr\1/' > {output}"

##dbSNP hg19
rule dbsnp_hg19_download:
    output: config["ref_dir"] + "/human/hg19/snp/dbsnp142.vcf"
    shell: "wget -q -O - ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b142_GRCh37p13/VCF/All_20150217.vcf.gz | gunzip -cd | sed 's/^\([^#]\)/chr\1/' > {output}"

## HapMap
# not included by now because the latest release is based on hg36

## 1K Genomes

## Mills

## Omni

######### CHAINS

rule download_hg19_to_hg38:
    output: config["ref_dir"] + "/human/hg37/chains/hg19ToHg38.over.chain"
    shell: "wget -q -O - http://hgdownload.cse.ucsc.edu/goldenPath/hg19/liftOver/hg19ToHg38.over.chain.gz | gunzip -cd > {output}"

######### OTHER DATABASES

## miRBase

MIRBASE_URL = "ftp://mirbase.org/pub/mirbase/CURRENT"

rule download_mirbase_mature:
    output: config["ref_dir"] + "/all/mirbase/mature.fa"
    shell: "wget -q -O - {MIRBASE_URL}/mature.fa.gz | gunzip > {output}"

rule generate_mirbase_mature_human:
    input: config["ref_dir"] + "/all/mirbase/mature.fa"
    output: config["ref_dir"] + "/human/hg38/mirbase/mature.fa"
    shell: "grep -A 1 '^>hsa' {input} | grep -v '^-' | sed 's/U/T/g' > {output}"

rule download_mirbase_hairpin:
    output: config["ref_dir"] + "/all/mirbase/hairpin.fa"
    shell: "wget -q -O - {MIRBASE_URL}/hairpin.fa.gz | gunzip > {output}"

rule generate_mirbase_hairpin_human:
    input: config["ref_dir"] + "/all/mirbase/hairpin.fa"
    output: config["ref_dir"] + "/human/hg38/mirbase/hairpin.fa"
    shell: "grep -A 1 '^>hsa' {input} | grep -v '^-' | sed 's/U/T/g' > {output}"
