# Snakefile for non-Illumina Amplicon-Kit workflow

configfile: "config.json"
include: "/data/DefaultPipelines/rules/references/human.rules"
include: "/data/DefaultPipelines/rules/references/illumina.rules"
include: "/data/DefaultPipelines/rules/trim/trimmomatic.rules"
include: "/data/DefaultPipelines/rules/mapping/bwa_mem_pe.rules"
include: "/data/DefaultPipelines/rules/mapping/samfiles.rules"
include: "/data/DefaultPipelines/rules/mapping/picard.rules"
include: "/data/DefaultPipelines/rules/coverage/bed_coverage.rules"
include: "/data/DefaultPipelines/rules/coverage/plots.rules"
include: "/data/DefaultPipelines/rules/variant/gatk.rules"
include: "/data/DefaultPipelines/rules/qa/fastqc.rules"
include: "/data/DefaultPipelines/rules/de/feature_count.rules"

rule all:
	input:
		"plots/coverage/bwa/hg19/quality_trimmed/BRCA1-ENST00000357654.pdf"
		,"plots/CNV/bwa/hg19/quality_trimmed/BRCA1-ENST00000357654.pdf"
		,"plots/coverage/bwa/hg19/quality_trimmed/Total.pdf"
		,expand("results/coverage/bwa/hg19/quality_trimmed/{patient}_20_20.lower100", patient = config["samples"].keys() )
		,expand("results/qa/fastqc/quality_trimmed/{unit}_fastqc.html", unit = [w.replace('.fastq.gz', '') for w in sum(config["units"].values(), [])] )
		,expand("results/variants/gatk/hg19/quality_trimmed/{patient}.vcf", patient = config["samples"].keys() )
		,"results/variants/gatk/hg19/quality_trimmed/all.vcf"
