#############################################################################
#          Script to generate human reference data.                         #
#                                                                           #
# Date: 23.02.2015                                                          #
#                                                                           #
# Example usage: snakemake --snakefile human.rules                          #
#############################################################################

DATA_DIR = "/data/references/"

######### GENOMES

## genome hg 38

# download genome

# build bwa index

## genome hg 37

# download genome

# build bwa index

######### VARIANT DATABASES

## dbSNP
rule get_dbsnp:
    output: DATA_DIR + "human/hg38/snp/dbsnp.vcf"
    shell: "wget -O - ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b142_GRCh38/VCF/All_20150114.vcf.gz | gunzip -cd > {output}"

## HapMap
# not included by now because the latest release is based on hg36

## 1K Genomes

## Mills

## Omni



######### OTHER DATABASES

## miRBase

MIRBASE_URL = "ftp://mirbase.org/pub/mirbase/CURRENT"

rule download_mirbase:
    output: DATA_DIR + "all/mirbase/mature.fa"
    shell: "wget -q -O -  {MIRBASE_URL}/mature.fa.gz | gunzip > {output}"

rule extract_human_mirbase:
    input: DATA_DIR + "all/mirbase/mature.fa"
    output: "human/hg38/mirbase/human_mature.fa"
    shell: "grep -A 1 Homo {input} | sed '/^--$/d' > {output}"

