#############################################################################
#          Script to generate all reference data.                           #
#                                                                           #
# Date: 02.01.2015                                                          #
#                                                                           #
# Example usage: snakemake --snakefile project.snake                        #


#############################################################################
#                              HUMAN DATA SETS                              #
#############################################################################

######### GENOMES

## genome hg 38

# download genome

# build bwa index

## genome hg 37

# download genome

# build bwa index

######### VARIANT DATABASES

## dbSNP
rule get_dbsnp:
    output: "human/hg38/snp/dbsnp.vcf"
    shell: "wget -O - ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b142_GRCh38/VCF/All_20150114.vcf.gz | gunzip -cd > {output}"

## HapMap
# not included by now because the latest release is based on hg36

## 1K Genomes

## Mills

## Omni



######### OTHER DATABASES

## miRBase

MIRBASE_URL = "ftp://mirbase.org/pub/mirbase/CURRENT"

rule download_mirbase:
    output: "all/mirbase/mature.fa"
    shell: "wget -q -O -  {MIRBASE_URL}/mature.fa.gz | gunzip > {output}"

rule extract_human_mirbase:
    input: "all/mirbase/mature.fa"
    output: "human/hg38/mirbase/human_mature.fa"
    shell: "grep -A 1 Homo {input} | sed '/^--$/d' > {output}"

### ILLUMINA DATA SETES

ILLUMINA_URL = "http://support.illumina.com/content/dam/illumina-support/documents"

## KITS

# trusight one kit

rule one_manifest_download:
    """Download trusight one manifest."""
    output: "illumina/kits/trusight-one-bed-may-2014.bed"
    shell: "wget -q -O - {ILLUMINA_URL}/downloads/productfiles/trusight/trusight-one-bed-may-2014.zip | gunzip > {output}"

# trusight cancer kit

rule cancer_manifest_download:
    """Download trusight cancer kit manifest"""
    output: "illumina/kits/trusight_cancer_manifest_a.bed"
    shell: "wget -q -O {output} {ILLUMINA_URL}/documentation/chemistry_documentation/trusight/trusight_cancer_manifest_a.bed"

rule cancer_manifest_pretify:
    """Pretifies the manifest file so it is easily readable by R (ggplot2)"""
    input: "illumina/kits/trusight_cancer_manifest_a.bed"
    output: "illumina/kits/trusight_cancer_manifest_a_pret.bed"
    shell: "cat {input} | "
           "sed 's/chr//' | " # trim first 'chr' in illumina manifest
           "sed 's/\.chr[1-9,X,Y]\+//' | " # remove 'chr{.+}' from location
           "sed 's/\./\t/' | " # replace first '.' in location to a TAB
           "sed 's/\./-/' > {output}" # replace last '.' in location to a '-' to pretify the region

rule cancer_manifest_extend_exons:
    """Extends the start/end of each feature by -20/+20 bp"""
    input: "illumina/kits/trusight_cancer_manifest_a_pret.bed"
    output: "illumina/kits/trusight_cancer_manifest_a_pret_ext.bed"
    shell: "awk '{{print $1\"\t\"$2-20\"\t\"$3+20\"\t\"$4\"\t\"$5}}' {input} > {output}" # add -20/+20 to column 2/3

#############################################################################
#                              BARLEY DATA SETS                             #
#############################################################################

rule barley_genome:
    output: "barley/morex/genome/morex.fa"
    shell: "wget -q -O - ftp://ftpmips.helmholtz-muenchen.de/plants/barley/public_data/sequences/assembly3_WGSMorex_renamed_blastable_carma.zip | gunzip > {output}"

rule barley_hc_genes:
    """Download high confidence gene set."""
    output: "barley/morex/genes/highconf.gtf"
    shell: "wget -q -O - ftp://ftpmips.helmholtz-muenchen.de/plants/barley/public_data/genes/barley_HighConf_genes_MIPS_21Aug12_Transcript_and_CDS_structure.gtf | sed 's/^contig/morex_contig/' > {output}"

rule barley_lc_genes:
    """Download low confidence gene set."""
    output: "barley/morex/genes/lowconf.gtf"
    shell: "wget -q -O - ftp://ftpmips.helmholtz-muenchen.de/plants/barley/public_data/genes/barley_LowConf_genes_MIPS_21Aug12_Transcript_and_CDS_structure.gtf | sed 's/^contig/morex_contig/' > {output}"

rule barley_genes:
    """Merge high and low confidence genes."""
    input: hc = "barley/morex/genes/highconf.gtf",
           lc = "barley/morex/genes/lowconf.gtf"
    output: "barley/morex/genes/all.gtf"
    shell: "cat {input.hc} {input.lc} > {output}"
           
