#############################################################################
#          Rules for counting.                                              #
#                                                                           #
# Date: 23.06.2015                                                          #
#                                                                           #
#############################################################################

rule feature_count:
    """Assigns counts to annotated features using name sorted BAMs."""
    input: bams = lambda wildcards: expand("results/mapping/" + wildcards.mapper + "/" + wildcards.ref + "/" + wildcards.kind + "/{unit}.bam", unit = config["units"].keys()),
           gtf = config["ref_dir"] + "/human/{ref}/genes/all.gtf",
    output: "results/de/featureCounts/{mapper}/{ref}/{kind}/all.counts"
    threads: 4
    shell: 'featureCounts -T {threads} --primary --minOverlap 10 -t "CDS" -g "gene_id" -a {input.gtf} -o {output} {input.bams}'
    
# # not tested until now
# rule count_mapped_sequences:
#     """Counts how often a sequence was mapped by reads."""
#     input: bams = expand("results/mapping/{{reference}}/{{kind}}/{unit}_name.bam", unit = config["units"].keys() ) ,
#     output: "results/de/{reference}/{kind}/{unit}.counts"
#     shell: "samtools view -F 4 {input} | cut -f 3 | sort | uniq -c | sed 's/^[ \t]*//;s/[ \t]*$//' > {output}"
